### ISSUE TITLE
 sh: cross-env: Permission denied in Gitlab pipeline (local build)

### ISSUE THREAD

**Issue by RosiersRobin** [lorisleiva/laravel-deployer#126]
Hellow

So I have the building in the pipeline etc... working, however, when I try now to run the npm run production, however I get the following error:

```
➤ Executing task npm:install
[localhost] > if [ -d /builds/rallypodium-reporting/rpr-site/vendor/lorisleiva/laravel-deployer/src/strategy/../../../../../node_modules ]; then echo 'true'; fi
[localhost] < true
[localhost] > cp -R /builds/rallypodium-reporting/rpr-site/vendor/lorisleiva/laravel-deployer/src/strategy/../../../../../node_modules /tmp/rpr/releases/1
[localhost] > which npm
[localhost] < /usr/bin/npm
[localhost] > cd /tmp/rpr/releases/1 && /usr/bin/npm install
[localhost] < npm WARN
[localhost] < optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/fsevents):
[localhost] < npm
[localhost] <  WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
[localhost] < audited 17261 packages in 14.219s
[localhost] < found 0 vulnerabilities
• done on [localhost]
✔ Ok [23s 915ms]
➤ Executing task npm:production
[localhost] > cd /tmp/rpr/releases/1 && (/usr/bin/npm run production)
[localhost] < > @ production /tmp/rpr/releases/1
[localhost] < > cross-env NODE_ENV=production node_modules/webpack/bin/webpack.js --no-progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js
[localhost] < sh: cross-env: Permission denied
[localhost] < npm
[localhost] <  ERR! code ELIFECYCLE
[localhost] < npm
[localhost] < ERR! errno 126
[localhost] < npm
[localhost] < ERR! @ production: `cross-env NODE_ENV=production node_modules/webpack/bin/webpack.js --no-progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js`
[localhost] < npm ERR!
[localhost] <  Exit status 126
[localhost] < npm ERR! 
[localhost] < npm
[localhost] <  ERR! Failed at the @ production script.
[localhost] < npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[localhost] < npm ERR!
[localhost] <  A complete log of this run can be found in:
[localhost] < npm ERR!     /root/.npm/_logs/2019-07-02T21_28_10_198Z-debug.log
➤ Executing task deploy:failed
```

This is my gitlab VI file:

```
image: lorisleiva/laravel-docker:latest

.init_staging_ssh: &init_staging_ssh |
  eval $(ssh-agent -s)
  echo "$SSH_PRIVATE_KEY_STAGING" | tr -d '\r' | ssh-add - > /dev/null
  mkdir -p ~/.ssh
  chmod 700 ~/.ssh
  [[ -f /.dockerenv ]] && echo -e "Host *\n\tStrictHostKeyChecking no\n\n" > ~/.ssh/config

.init_production_ssh: &init_production_ssh |
  eval $(ssh-agent -s)
  echo "$SSH_PRIVATE_KEY_STAGING" | tr -d '\r' | ssh-add - > /dev/null
  mkdir -p ~/.ssh
  chmod 700 ~/.ssh
  [[ -f /.dockerenv ]] && echo -e "Host *\n\tStrictHostKeyChecking no\n\n" > ~/.ssh/config

.change_file_permissions: &change_file_permissions |
  find . -type f -not -path "./vendor/*" -exec chmod 664 {} \;
  find . -type d -not -path "./vendor/*" -exec chmod 775 {} \;

cache:
  untracked: true
  key: "$CI_COMMIT_REF_NAME"
  paths:
    - vendor/
    - node_modules/

stages:
  - preparation
  - building
  - deploy-build
  - deploy

variables:
  CONTAINER_MYSQL_IMAGE: mysql:5.7
  CONTAINER_NODE_IMAGE: node:lts
  MYSQL_DATABASE: rpr
  MYSQL_ROOT_PASSWORD: toor
  DOCKER_DRIVER: overlay


##########################
## INSTALL DEPENDENCIES ##
##########################
composer:
  stage: preparation
  services:
    - $CONTAINER_MYSQL_IMAGE
  before_script:
    - php -v # Have PHP version in logs
    - composer --version # Have Composer version in logs
    - git --version # Have Git version in logs
    - ls -lah # Overview of folders
    - cp -pRf .env.example .env
  script:
    - composer install --no-ansi --dev --no-interaction --no-progress --no-scripts --optimize-autoloader
  after_script:
    - php artisan key:generate
  artifacts:
    expire_in: 1hr
    paths:
      - ./.env
      - ./vendor
      - ./public/bundles

node:
  stage: preparation
  before_script:
    - npm install --save-dev cross-env
    - npm install
  script:
    - npm run production
  artifacts:
    expire_in: 1hr
    paths:
      - ./node_modules

########################
## BUILD DEPENDENCIES ##
########################
build-assets:
  stage: building
  dependencies:
    - composer
    - node
  script:
    - npm run production
  artifacts:
    paths:
      - public/css/
      - public/js/
      - public/fonts/
      - public/admin
      - public/images
      - public/svg
      - public/mix-manifest.json
    expire_in: 1 days
    when: always

db-seeding:
  stage: building
  services:
    - $CONTAINER_MYSQL_IMAGE
  dependencies:
    - composer
    - node
  script:
    - php artisan migrate:fresh --seed
  artifacts:
    paths:
      - ./storage/logs # for debugging
    expire_in: 1 days
    when: on_failure

#####################
## DEPLOY THE CODE ##
#####################

# Staging deploy
staging:
  stage: deploy
  environment:
    name: staging
    url: https://staging.rallypodium.be
  before_script:
    - docker-php-ext-install bcmath
    - docker-php-ext-install exif
    - *init_staging_ssh
    - *change_file_permissions
  script:
    - php artisan deploy staging -vvv
  only:
    - development

```

This is my deploy config:

```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Default deployment strategy
    |--------------------------------------------------------------------------
    |
    | This option defines which deployment strategy to use by default on all
    | of your hosts. Laravel Deployer provides some strategies out-of-box
    | for you to choose from explained in detail in the documentation.
    |
    | Supported: 'basic', 'firstdeploy', 'local', 'pull'.
    |
    */

    'default' => 'local',

    /*
    |--------------------------------------------------------------------------
    | Custom deployment strategies
    |--------------------------------------------------------------------------
    |
    | Here, you can easily set up new custom strategies as a list of tasks.
    | Any key of this array are supported in the `default` option above.
    | Any key matching Laravel Deployer's strategies overrides them.
    |
    */

    'strategies' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Hooks
    |--------------------------------------------------------------------------
    |
    | Hooks let you customize your deployments conveniently by pushing tasks 
    | into strategic places of your deployment flow. Each of the official
    | strategies invoke hooks in different ways to implement their logic.
    |
    */

    'hooks' => [
        // Right before we start deploying.
        'start' => [
            //
        ],
        
        // Code and composer vendors are ready but nothing is built.
        'build' => [
            'npm:install',
            'npm:production',
        ],
        
        // Deployment is done but not live yet (before symlink)
        'ready' => [
            'artisan:storage:link',
            'artisan:view:clear',
            'artisan:cache:clear',
            'artisan:migrate',
        ],
        
        // Deployment is done and live
        'done' => [
            'artisan:cache:clear',
            'artisan:config:clear',
            'artisan:view:clear',
            'fpm:reload',
        ],
        
        // Deployment succeeded.
        'success' => [
            'artisan:cache:reload',
            'artisan:queue:restart',
        ],
        
        // Deployment failed.
        'fail' => [
            //
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Deployment options
    |--------------------------------------------------------------------------
    |
    | Options follow a simple key/value structure and are used within tasks
    | to make them more configurable and reusable. You can use options to
    | configure existing tasks or to use within your own custom tasks.
    |
    */

    'options' => [
        'application' => env('APP_NAME'),
        'repository' => 'git@gitlab.com:rallypodium-reporting/rpr-site.git',
        'default_timeout' => null,
        'local_deploy_path' => '/tmp/rpr',
        'local_upload_options' => [
            'options' => [
                '--exclude=".env.example"',
                '--exclude=".git/"',
                '--exclude=".gitattributes"',
                '--exclude=".gitignore"',
                '--exclude=".gitlab-ci.yml"',
                '--exclude="composer.json"',
                '--exclude="composer.lock"',
                '--exclude="Dockerfile"',
                '--exclude="node_modules/"',
                '--exclude="package-lock.json"',
                '--exclude="package.json"',
                '--exclude="phpunit.xml"',
                '--exclude="server.php"',
                '--exclude="storage/"',
                '--exclude="tests/"',
                '--exclude="vendor/"',
                '--exclude="webpack.mix.js"',
            ],
        ],
        'php_fpm_service' => 'php7.3-fpm',
    ],

    /*
    |--------------------------------------------------------------------------
    | Hosts
    |--------------------------------------------------------------------------
    |
    | Here, you can define any domain or subdomain you want to deploy to.
    | You can provide them with roles and stages to filter them during
    | deployment. Read more about how to configure them in the docs.
    |
    */

    'hosts' => [
        'rallypodium.be' => [
            'deploy_path' => '/var/www/rallypodium.be',
            'user' => 'robin',
            'stage' => 'production',
            'branch' => 'master'
        ],
        '213.118.218.134' => [
            'deploy_path' => '/var/www/staging.rallypodium.be',
            'user' => 'root',
            'stage' => 'staging',
            'branch' => 'development',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Localhost
    |--------------------------------------------------------------------------
    |
    | This localhost option give you the ability to deploy directly on your
    | local machine, without needing any SSH connection. You can use the
    | same configurations used by hosts to configure your localhost.
    |
    */

    'localhost' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Include additional Deployer recipes
    |--------------------------------------------------------------------------
    |
    | Here, you can add any third party recipes to provide additional tasks, 
    | options and strategies. Therefore, it also allows you to create and
    | include your own recipes to define more complex deployment flows.
    |
    */

    'include' => [
        'deployer/recipe/artisan_refresh_cache.php'
    ],

    /*
    |--------------------------------------------------------------------------
    | Use a custom Deployer file
    |--------------------------------------------------------------------------
    |
    | If you know what you are doing and want to take complete control over
    | Deployer's file, you can provide its path here. Note that, without
    | this configuration file, the root's deployer file will be used.
    |
    */

    'custom_deployer_file' => false,

];
```

**Comment by Turpoint** [comment]
I have the same error. Were you able to fix it?

**Comment by lorisleiva** [comment]
This issue seems to be tightly coupled to the environment of your hosts. Feel free to open a new issue with more elaborated information.

**Comment by olavorn** [comment]
We are having the same issue. Gitlab Runner on a docker container.
